### PR TITLE
CHAT-383 Improve user suggestion when create new chat room

### PR DESCRIPTION
**Pull request by nttuyen**

### CODE REVIEW

File: server/src/main/java/org/exoplatform/chat/services/mongodb/MongoModule.java [hunk 1]
```diff
@@ -15,5 +19,21 @@ protected void configure() {
...
+      public T get() {
+        RootContainer rootContainer = RootContainer.getInstance();
+        T ret = rootContainer.getComponentInstanceOfType(clazz);
+        if(ret == null) {
+          PortalContainer portalContainer = PortalContainer.getInstance();
+          ret = portalContainer.getComponentInstanceOfType(clazz);
+        }
+        return ret;
+      }
+    };
```
Comment (trongtt) [inline server/src/main/java/org/exoplatform/chat/services/mongodb/MongoModule.java]: IdentityManager is only used in UserServiceImpl.java, so just move these code there for now.

File: server/src/main/java/org/exoplatform/chat/services/mongodb/UserServiceImpl.java [hunk 2]
```diff
@@ -493,6 +504,121 @@ private SpaceBean getSpace(String roomId, String dbName)
...
+              }
+            }
+          } catch (Exception ex) {
+            LOG.info(ex.getMessage());
+          }
+          //
+          remain = limit - users.size();
+        }
+      }
+    }
```
Comment (trongtt) [inline server/src/main/java/org/exoplatform/chat/services/mongodb/UserServiceImpl.java:587]: This code block of finding connections by user info should be moved to Social service, even we may need to introduce a new method.

It helps support another social RDMBS implementation add-on works well later.

File: server/src/main/java/org/exoplatform/chat/services/mongodb/UserServiceImpl.java [hunk 3]
```diff
@@ -493,6 +504,121 @@ private SpaceBean getSpace(String roomId, String dbName)
     return users;
   }
 
+  @Override
+  public List<UserBean> getSuggestionUsers(String currentUserName, String roomId, String filter, String dbName, int limit) {
```
Comment (thomasdelhomenie) [inline server/src/main/java/org/exoplatform/chat/services/mongodb/UserServiceImpl.java:512]: I am not sure to understand the add of the roomId here. We want to suggest users when adding members in a team, so why we would suggest someone already in the team ?
The spec mentions that "In Chat add-on, when creating a team, the autocompletion to add people should list my connections first (1) and then other users (2)."
I should miss something here :)

File: application/src/main/webapp/js/chat.js [hunk 4]
```diff
@@ -1648,17 +1659,41 @@ ChatApplication.prototype.getUsers = function(roomId, callback, asString) {
...
+  // Cache result
+  var _this = this;
+  if (this.getAllUsersCache == undefined) {
+    this.getAllUsersCache = {};
+  }
+  var cacheKey = filter;
+  if (isSearchInRoomFirst) {
+    cacheKey += '__' + this.targetUser;
+  }
+  if (this.getAllUsersCache[cacheKey] != undefined) {
```
Comment (thomasdelhomenie) [inline application/src/main/webapp/js/chat.js:1708]: Adding a cache here could be a good idea, but it seems it is probably too basic. Once the cache has been populated, it is never invalidated, so if a new user has been created or created in the meanwhile, the cache will not be updated. The only way to reload the cache is to refresh the page.
Always, there is no max capacity in this cache.

File: server/src/main/java/org/exoplatform/chat/services/mongodb/UserServiceImpl.java (hunk 3 above)
Comment (nttuyen) [inline server/src/main/java/org/exoplatform/chat/services/mongodb/UserServiceImpl.java:512]: Yes, there is one more rule in assigning a task in chat application:

> In Chat add-on, when assigning a task to someone, the autocompletion of the field "Assign to" should list the member of the team first (1) and then my connections (2), then other users (3)

That's why I have to add "roomId" here.

Comment (thomasdelhomenie) [inline server/src/main/java/org/exoplatform/chat/services/mongodb/UserServiceImpl.java:512]: I knew I missed something :) Thanks @nttuyen

File: server/src/main/java/org/exoplatform/chat/services/mongodb/UserServiceImpl.java [hunk 5]
```diff
@@ -493,6 +504,121 @@ private SpaceBean getSpace(String roomId, String dbName)
...
+      searchByQuery(users, dbName, query, limit);
+    }
+
+    int remain = limit - users.size();
+
+    // Then search in connection
+    if (remain > 0 && currentUserName != null && !"".equals(currentUserName)) {
+      // We can not inject IdentityManager to this service here because it threw NPE,
+      // maybe Guice container only return this service instance when portal container started
+      IdentityManager identityManager = GuiceManager.getInstance().getInstance(IdentityManager.class);
```
Comment (thomasdelhomenie) [inline server/src/main/java/org/exoplatform/chat/services/mongodb/UserServiceImpl.java]: How did you try to inject it ?
I think we should open a dedicated issue if there is no way to inject a service cleanly. It is ugly to have to use this solution.

Comment (nttuyen) [inline server/src/main/java/org/exoplatform/chat/services/mongodb/UserServiceImpl.java]: I tried to inject by using annotation @Inject but maybe chatServer.war is not in portal context and NPE is thrown.
Anyway, I switched to use CommonsUtils for getting service from exo container as other places. I think it's easier to understand than the way we have to configure provider than get service via GuiceManager directly.

File: server/src/main/java/org/exoplatform/chat/services/mongodb/UserServiceImpl.java (hunk 2 above)
Comment (nttuyen) [inline server/src/main/java/org/exoplatform/chat/services/mongodb/UserServiceImpl.java:587]: I improved by using the method #getConnectionsByFilter of RelationshipManager, it works same as what we did but it allows us to re-implement on social-rdbms addon to get a better performance.

File: application/src/main/webapp/js/chat.js (hunk 4 above)
Comment (nttuyen) [inline application/src/main/webapp/js/chat.js:1708]: IMO, refresh the page to clear cache in JavaScript is enough for us now. Usually, user does not do this search frequently and the size of cache is not too big. 
Actually, some javascript libraries in PLF work in the same way (as mentioning and assign/coworker in task management)

Comment (thomasdelhomenie) [inline application/src/main/webapp/js/chat.js:1708]: Having to refresh the page to see the last data is not what I expect from a Single Page Application. My chat is always open in a browser tab and I never refresh it.
That said, I agree the cases I mentionned are corner cases, so we can go like that for now, but I think we should study the use of a more advanced client cache (home-made or ext lib). WDYT ?

Comment (nttuyen) [inline application/src/main/webapp/js/chat.js:1708]: I agree, we should consider to use a more advanced client-cache if we really need in other issue. 
I will create an issue to take note, ok?

Comment (thomasdelhomenie) [inline application/src/main/webapp/js/chat.js:1708]: yep :+1:

File: server/src/main/java/org/exoplatform/chat/services/mongodb/MongoModule.java [hunk 6]
```diff
@@ -1,10 +1,14 @@
 package org.exoplatform.chat.services.mongodb;
 
 import com.google.inject.AbstractModule;
+import com.google.inject.Provider;
```
Comment (thomasdelhomenie) [inline server/src/main/java/org/exoplatform/chat/services/mongodb/MongoModule.java]: Useless imports

File: server/src/main/java/org/exoplatform/chat/server/ChatServer.java [hunk 7]
```diff
@@ -869,7 +863,8 @@ else if (usersNew.contains(userExist))
     }
     else
     {
-      users = userService.getUsers(filter, true, dbName);
+      //users = userService.getUsers(filter, true, dbName);
```
Comment (thomasdelhomenie) [inline server/src/main/java/org/exoplatform/chat/server/ChatServer.java]: to remove

File: application/src/main/webapp/js/chat.js [hunk 8]
```diff
@@ -830,51 +832,60 @@ var chatApplication = new ChatApplication();
...
-          var users = users();
-          var $userResults = $("."+prefix+"-users-results");
-          $userResults.css("display", "none");
-          var html = "";
-          if (!withCurrentUser) {
+      } else if (filter != searchingValue){
+        searchingValue = filter;
+        if (searchTimeout !== false) {
+          clearTimeout(searchTimeout);
+          searchTimeout = false;
```
Comment (trongtt) [inline application/src/main/webapp/js/chat.js]: Seems set _searchTimeout = false_ is useless. We don't even need the condition check _searchTimeout !== false_, always clearTimeout() before set a new one.

**Comment by nttuyen** [comment]
We should not merge this PR, I create a replacement one to merge to develop branch at exo-addons: https://github.com/exo-addons/chat-application/pull/45

**Pull request closed without merging**